src/cljs/athens/self_hosted/client.cljs

```diff
@@ -0,0 +1,449 @@
+                                           (if (= :block/children a)
+                                             (conj (:block/children entity #{})
+                                                   (insert-id v))
+                                             v)))
+                                  {:db/id        (insert-id entity-id)
+                                   :remote/db-id entity-id}
+                                  datoms)))
+                 {})
+         ;; only entities
+         vals)))
+
+
+(defn- db-dump-handler
+  [last-tx {:keys [datoms] :as args}]
+  (js/console.debug "Received DB Dump")
+  (let [entities (reconstruct-entities-from-db-dump datoms)]
+    (js/console.debug "Reconstructed" (count entities) "entities")
+    (rf/dispatch [:reset-conn (d/empty-db db/schema)])
+    (rf/dispatch [:transact entities])
+    (rf/dispatch [:remote/last-seen-tx! last-tx])
+    (rf/dispatch [:db/sync])
+    (rf/dispatch [:remote/connected])
+    (js/console.log "✅ Transacted DB dump. last-seen-tx" last-tx)))
+
+
+(defn- presence-online-handler
+  [args]
+  (let [username (:username args)]
+    ;; TODO manage connected users in re-frame
+    (js/console.log "User online:" username)))
+
+
+(defn- server-event-handler
+  [{:event/keys [id last-tx type args] :as packet}]
+  (js/console.log "<-" id ", last-tx:" last-tx ", type:" type)
+  (js/console.debug "WSClient: server event:" (pr-str packet))
+  (if (schema/valid-server-event? packet)
+
+    (condp = type
+      :datascript/tx-log  (ds-tx-log-handler args)
+      :datascript/db-dump (db-dump-handler last-tx args)
+      :presence/online    (presence-online-handler args))
+
+    (js/console.warn "TODO invalid server event" (pr-str (schema/explain-server-event packet)))))
+
+
+(def ^:private datom-reader
+  (transit/read-handler
+    (fn [[e a v tx added]]
+      {:e     e
+       :a     a
+       :v     v
+       :tx    tx
+       :added added})))
+
+
+(defn- message-handler
+  [event]
+  (let [packet    (->> event
+                       .-data
+                       (transit/read
+                         (transit/reader
+                           :json
+                           {:handlers
+                            {:datom datom-reader}})))
+        {:event/keys
+         [id]}    packet
+        req-event (get @awaiting-response id)]
+    (js/console.log "message-handler" (pr-str packet))
+    (if req-event
+      (awaited-response-handler req-event packet)
+      (server-event-handler packet))))
+
+
+(defn- remove-listeners!
+  [connection]
+  (doto connection
+    (.removeEventListener "close" close-handler)
+    (.removeEventListener "message" message-handler)
+    (.removeEventListener "open" open-handler)))
+
+
+(defn- close-handler
+  [event]
+  (js/console.log "WSClient Disconnected:" event)
+  (let [connection (.-target event)
+        url        (.-url connection)]
+    (remove-listeners! connection)
+    (delayed-reconnect! url)))
```

Comment:
Why reconnect on close?

Reply:
`on-close` means we didn't try to stop the connection, it was unintentional like a network issue or closing laptop, so we should try to reconnect.

There is a specific close event that should only be called intentionally.